// express-vue/dist/utils/render.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.renderVueMixins = exports.renderVueComponents = exports.renderHtmlUtil = exports.layoutUtil = exports.renderUtil = undefined;

var _models = require('../models');

var _string = require('./string');

var _string2 = _interopRequireDefault(_string);

var _head = require('./head');

var _head2 = _interopRequireDefault(_head);

var _paramCase = require('param-case');

var _paramCase2 = _interopRequireDefault(_paramCase);

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderer = require('vue-server-renderer').createRenderer();

var appRegex = /{{{app}}}/igm;
var scriptRegex = /{{{script}}}/igm;
var headRegex = /<\/head>/igm;
var types = new _models.Types();

function createApp(script, defaults) {
    if (defaults.options.vue && defaults.options.vue.mixins) {
        if (defaults.options.vue.mixins.length > 0) {
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = defaults.options.vue.mixins[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var mixin = _step.value;

                    _vue2.default.mixin(mixin);
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }
        }
    }

    return new _vue2.default(script);
}

function layoutUtil(components) {
    var layout = {};
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
        for (var _iterator2 = components[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var component = _step2.value;

            switch (component.type) {
                case types.LAYOUT:
                    layout = component;
                    break;
                case types.COMPONENT:
                    layout.script = component.script;
                    break;
                case types.SUBCOMPONENT:
                    if (layout.script.components) {
                        layout.script.components[component.name] = component.script;
                    } else {
                        layout.script.components = {};
                        layout.script.components[component.name] = component.script;
                    }
                    break;
            }
        }
    } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
                _iterator2.return();
            }
        } finally {
            if (_didIteratorError2) {
                throw _iteratorError2;
            }
        }
    }

    return layout;
}

function renderUtil(layout, renderedScriptString, defaults) {
    return new Promise(function (resolve, reject) {
        var app = createApp(layout.script, defaults);
        renderer.renderToString(app, function (error, renderedHtml) {
            if (error) {
                reject(error);
            }
            var html = '';
            var head = '';
            html = layout.template.replace(appRegex, '<div id="app">' + renderedHtml + '</div>');
            html = html.replace(scriptRegex, renderedScriptString);
            head = (0, _head2.default)(defaults.options.vue);
            html = html.replace(headRegex, head);
            resolve(html);
        });
    });
}

function renderVueComponents(script, components) {
    var componentsString = '';
    for (var component in components) {
        if (components.hasOwnProperty(component)) {
            var currentComponent = components[component];
            if (currentComponent.type === types.SUBCOMPONENT) {
                componentsString = componentsString + ('Vue.component(\'' + (0, _paramCase2.default)(currentComponent.name) + '\', ' + (0, _string2.default)(currentComponent.script) + ');\n');
            }
        }
    }

    return componentsString;
}

function renderVueMixins(mixins) {
    var mixinString = '';
    var _iteratorNormalCompletion3 = true;
    var _didIteratorError3 = false;
    var _iteratorError3 = undefined;

    try {
        for (var _iterator3 = mixins[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            var mixin = _step3.value;

            mixinString = mixinString + ('Vue.mixin(' + (0, _string2.default)(mixin) + ');\n');
        }
    } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion3 && _iterator3.return) {
                _iterator3.return();
            }
        } finally {
            if (_didIteratorError3) {
                throw _iteratorError3;
            }
        }
    }

    return mixinString;
}

function renderedScript(script, components, mixins) {

    var componentsString = renderVueComponents(script, components);
    var mixinString = mixins !== undefined ? renderVueMixins(mixins) : '';
    var scriptString = (0, _string2.default)(script);
    var debugToolsString = '';

    if (process.env.VUE_DEV) {
        debugToolsString = 'Vue.config.devtools = true;';
    }
    return '<script>\n        (function () {\n            \'use strict\'\n            ' + mixinString + '\n            ' + componentsString + '\n            var createApp = function () {\n                return new Vue(\n                    ' + scriptString + '\n                )\n            }\n            if (typeof module !== \'undefined\' && module.exports) {\n                module.exports = createApp\n            } else {\n                this.app = createApp()\n            }\n        }).call(this)\n        ' + debugToolsString + '\n        app.$mount(\'#app\');\n    </script>';
}

function renderHtmlUtil(components, defaults) {
    var mixins = [];
    if (defaults.options.vue && defaults.options.vue.mixins) {
        mixins = defaults.options.vue.mixins;
    }
    var layout = layoutUtil(components);
    var renderedScriptString = renderedScript(layout.script, components, mixins);
    return renderUtil(layout, renderedScriptString, defaults);
}

exports.renderUtil = renderUtil;
exports.layoutUtil = layoutUtil;
exports.renderHtmlUtil = renderHtmlUtil;
exports.renderVueComponents = renderVueComponents;
exports.renderVueMixins = renderVueMixins;
